app/javascript/extensible-components/index.ts

```diff
@@ -0,0 +1,69 @@
+import { IExtensionComponent, IMiQApiCallback } from './lib';
+
+const source = new Rx.Subject();
+const items = new Set();
+
+/**
+ * Class for easy creation of extensible component.
+ */
+export class ExtensibleComponent {
+  public unsubscribe: () => void;
+  constructor(public name: string, public api: IMiQApiCallback, public render: IMiQApiCallback){}
+}
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
+ */
+function addComponent(name: string, api?: IMiQApiCallback, render?: IMiQApiCallback) {
+  let newCmp = new ExtensibleComponent(name, api, render);
+  source.onNext({action: 'add', payload: newCmp});
+  return newCmp;
+}
+
+/**
+ * Helper function to subscribe to extension items based on component's name.
+ * @param cmpName name of component we want to subscribe to.
+ * @return object which has with and unsubscribe property. With is for callback to use found component and delete to
+ * unsubscribe from rxjs subject.
+ */
+function subscribe(cmpName: string) {
+  let unsubscribe;
+  return {
+    with: (callback) => {
+      unsubscribe = source
+        .map(sourceAction => sourceAction.action === 'add' ? sourceAction.payload : {})
+        .filter(component => component && component.name === cmpName)
+        .subscribe(cmp => cmp && callback(cmp));
+
+      ManageIQ.extensions.items.forEach((component) => {
+        component.name === cmpName && callback(component);
+      });
+    },
+    delete: () => unsubscribe && unsubscribe()
+  }
+}
+
+const extensions = {
+  addComponent: addComponent,
+  subscribe: subscribe,
+  getItems: () => items
+}
+
+ManageIQ.extensions = ManageIQ.extensions || extensions;
```

Comment:
I wonder if we should use webpack import (e.g.)
```js
import {subscribe, addComponent} from 'extensible-component';
```

IMHO we should expose it as a global var, but going forward I think its better to leverage webpack as you would get compile errors and live reloading when needed.

Reply:
The idea was that MiQ APIs are primarily exposed through globals, e.g. `window.ManageIQ.stuff`.

Given that, we also defined client libraries like `lib.ts` to be compiled into `lib.js` and published to npm registry, so that client JS code can do `import { addComponent } from 'miq-lib'`.

(I'd prefer those client libraries to be named `client-extension` and `client-redux`, instead of `lib`, but it's not a big deal.)

Reply:
@vojtechszocs again, I'm not against exposing those API's but legacy / others, but I wonder if its better to use import / export with webpack vs relay on window (for example for testing, live reload functionality etc).

Reply:
@ohadlevy normally you'd not use the globals, you'd just use `import { addComponent } from 'miq-lib'`. This applies to both core- and plugin- JS code.

Reply:
The idea behind this is that both redux and extensioble component will npm packages, pulled either from rails plugin or just included in `ui-classic`. So importing will look like `import { addComponent } from '@manageIq/extension'`. However it's convenient to have this in `ui-classic` the way it is now, so we don't have to rely on npm package and such, so for now it's easier to develop this way and move it to npm package later on.

About the global variables, that's mostly just for testing/legacy reasons. Plus when bundling webpack packages right now we'd duplicate a lot of code. So let's say we have 2 plugins both are using `Rx.js`, webpack will then create packs which are nearly 2 MB big, both packs will have same rx included in them.

Reply:
that should not be the case, webpack should be able to chunk it and have only one copy of it. have a look at https://webpack.js.org/plugins/commons-chunk-plugin/

Reply:
> Plus when bundling webpack packages right now we'd duplicate a lot of code.

MiQ Redux & extensible component infras need to live somewhere at runtime. For now, they're living in `ManageIQ.redux` and `ManageIQ.extensions` globals. As @karelhala wrote, without those infras living somewhere, we'd be duplicating the infra code in every client that uses it.

Reply:
@vojtechszocs while that is fine, when a new webpack owned js will import it - e.g.
```js
import { store } from 'miq/redux'
```
that would not duplicate anything at build or runtime... so what am I missing? :)

Reply:
> webpack should be able to chunk it and have only one copy of it

Using globals is most flexible and future-proof 😃 plus, any additional code lazy loaded & executed at runtime can easily interact with a global. Having a single global (namespace) representing the application, like `ManageIQ`, is a common practice.

Reply:
@ohadlevy The bit you're missing is that not all the code goes through webpack, and won't for a while.

Common chunkgs plugin would help - but has no effect at all until we stop using the Rails asset pipeline. And unless we spend the next month or two moving *everything* over to webpack, we will be using the asset pipeline for a while yet.

(Also, importing stuff can happen in new code, but.. not necessarily in the old one.)

Reply:
thanks @himdel I think we are both agreeing on all of your points.

I think what I'm saying, is that within webpack managed code, we should default to import vs using the globals.

Reply:
@ohadlevy the biggest problem here is that both redux and extensible component code is not bundled as [a library](http://webpack.github.io/docs/configuration.html#output-library). They sit in ui-classic for easier development and are treated as general packs. But once we have proper setup of these two libraries (plus commons chunk plugin set up in ui-components) and they'll be npm packages we (and any plugin using webpack) can import them as you said.

Reply:
> we should default to import vs using the globals

Agreed on that. We use the `ManageIQ` global variable as a place to store and reference the newly added mechanisms for any JS code out there. In ES6/webpack compatible code, we always prefer `import` over direct global access.

Reply:
@karelhala why is it different if its a npm package or not? as long as its being exported (via an npm package or on ui-classic/webpack) it should work just the same.

ideally, going forward, new code will land in webpack control, so it shouldnt be really used that much (e.g. using globals).

when you actually move it to a library, all you would need to change is your import line, e.g.
```js 
import { something } from './some-pack'
```
to
```js 
import { something } from 'some-package'
```

Reply:
I think we are in circles and talking about same things.

Yes if we are in webpack managed file we will favor webpack imports over global access variables. That's why we have lib files here - you will include addNewComponent which will reference to global variable, and once we move all files to webpack managed world we can just change the lib files in a way that they will use imported functions as wel.